File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

struct xy{
    int x;
    int y;
};
int dx[] = {0,0,1,-1};
int dy[] = {1,-1,0,0};

int visit[101][101];

vector<int> solution(vector<string> maps) {
    vector<int> answer;
    
    queue<xy> q;
    for(int i=0;i<maps.size();i++){
        for(int j=0;j<maps[0].size();j++){
            if(maps[i][j] == 'X' or visit[i][j])
                continue;
            int sum = 0;
            visit[i][j] = 1;
            q.push({i,j});
            
            while(q.size() !=0){
                int x = q.front().x;
                int y = q.front().y;
                sum += maps[x][y] - '0';
               
                q.pop();
                
                for(int k=0;k<4;k++){
                    int nx = x+dx[k];
                    int ny = y+dy[k];
                    if(nx >= maps.size() or nx < 0 or ny >= maps[0].size() or ny < 0)
                        continue;
                    if(maps[nx][ny] == 'X' or visit[nx][ny])
                        continue;
                    q.push({nx,ny});
                    visit[nx][ny] = 1;
                }
            }
            answer.push_back(sum);
        }
    }
    if(answer.size()==0)
        answer.push_back(-1);
    sort(answer.begin(),answer.end());
    
    return answer;
}

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.io.*;
import java.util.*;
class Solution {
    static char[][]map;
    static int[][]visit;
    static int cnt, n, m;
    public int[] solution(String[] maps) {
        int[] answer = {};
        
        ArrayList<Integer> an = new ArrayList<>();
        n = maps.length;
        m = maps[0].length();
        map = new char[n][m];
        visit = new int[n][m];
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++)
                map[i][j] = maps[i].charAt(j);
        }
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(map[i][j] == 'X') continue;
                if(visit[i][j] == 1) continue;
                dfs(i,j);
                an.add(cnt);
                cnt = 0;
            }
        }
        if(an.size() == 0) an.add(-1);
        answer = new int[an.size()];
        for(int i = 0; i < an.size(); i++)
            answer[i] = an.get(i);
        
        Arrays.sort(answer);
        return answer;
    }
    static void dfs(int x, int y){
        if(x < 0 || x >= n || y < 0 || y >= m) return;
        if(visit[x][y] == 1 || map[x][y] == 'X') return;
        
        
        visit[x][y] = 1;
        cnt += map[x][y] - '0';
        dfs(x + 1, y);
        dfs(x - 1, y);
        dfs(x, y + 1);
        dfs(x, y - 1);
    }
    
}